bundles/org.openhab.binding.renault/src/main/java/org/openhab/binding/renault/internal/renault/api/Car.java

```diff
@@ -0,0 +1,127 @@
+        }
+    }
+
+    public void setHVACStatus(JsonObject responseJson) {
+        try {
+            JsonObject attributes = getAttributes(responseJson);
+            if (attributes != null && attributes.get("hvacStatus") != null) {
+                hvacstatus = attributes.get("hvacStatus").getAsString().equals("on");
+            }
+        } catch (IllegalStateException | ClassCastException e) {
+            logger.warn("Error {} parsing HVAC Status: {}", e.getMessage(), responseJson);
+        }
+    }
+
+    public void setCockpit(JsonObject responseJson) {
+        try {
+            JsonObject attributes = getAttributes(responseJson);
+            if (attributes != null && attributes.get("totalMileage") != null) {
+                odometer = attributes.get("totalMileage").getAsDouble();
+            }
+        } catch (IllegalStateException | ClassCastException e) {
+            logger.warn("Error {} parsing Cockpit: {}", e.getMessage(), responseJson);
+        }
+    }
+
+    public void setLocation(JsonObject responseJson) {
+        try {
+            JsonObject attributes = getAttributes(responseJson);
+            if (attributes != null) {
+                if (attributes.get("gpsLatitude") != null) {
+                    gpsLatitude = attributes.get("gpsLatitude").getAsDouble();
+                }
+                if (attributes.get("gpsLongitude") != null) {
+                    gpsLongitude = attributes.get("gpsLongitude").getAsDouble();
+                }
+            }
+        } catch (IllegalStateException | ClassCastException e) {
+            logger.warn("Error {} parsing Location: {}", e.getMessage(), responseJson);
+        }
+    }
+
+    public void setDetails(JsonObject responseJson) {
+        try {
+            if (responseJson.get("assets") != null) {
+                JsonArray assetsJson = responseJson.get("assets").getAsJsonArray();
+                for (JsonElement asset : assetsJson) {
+                    if (asset.getAsJsonObject().get("assetType") != null
+                            && asset.getAsJsonObject().get("assetType").getAsString().equals("PICTURE")) {
+                        if (asset.getAsJsonObject().get("renditions") != null) {
+                            JsonArray renditions = asset.getAsJsonObject().get("renditions").getAsJsonArray();
+                            for (JsonElement rendition : renditions) {
+                                if (rendition.getAsJsonObject().get("resolutionType") != null
+                                        && rendition.getAsJsonObject().get("resolutionType").getAsString()
+                                                .equals("ONE_MYRENAULT_SMALL")) {
+                                    imageURL = rendition.getAsJsonObject().get("url").getAsString();
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                    if (!imageURL.isEmpty()) {
```

Comment:
There is a compilation warning here and your code is not good as imageURL could have been set previously.
Here is my proposal using a local variable:
```
                JsonArray assetsJson = responseJson.get("assets").getAsJsonArray();
                String url = null;
                for (JsonElement asset : assetsJson) {
                    if (asset.getAsJsonObject().get("assetType") != null
                            && asset.getAsJsonObject().get("assetType").getAsString().equals("PICTURE")) {
                        if (asset.getAsJsonObject().get("renditions") != null) {
                            JsonArray renditions = asset.getAsJsonObject().get("renditions").getAsJsonArray();
                            for (JsonElement rendition : renditions) {
                                if (rendition.getAsJsonObject().get("resolutionType") != null
                                        && rendition.getAsJsonObject().get("resolutionType").getAsString()
                                                .equals("ONE_MYRENAULT_SMALL")) {
                                    url = rendition.getAsJsonObject().get("url").getAsString();
                                    break;
                                }
                            }
                        }
                    }
                    if (url != null && !url.isEmpty()) {
                        imageURL = url;
                        break;
                    }
```

Reply:
There is a some data we could persist in memory and do not to waste electricity collecting again. Account ID car image url....  However I copied the python code and run through the collection steps ever 10 mins and discard the collected Session and Car.  This is simple and robust but inefficient.  Very happy to improve it but not sure how to do this in an openHAB compatible way.

What is the best way to store the collected data?  Can I get the channel and see if there is an image?  Can I store the account ID etc in memory in the Session and not discard it each run?

Or we just get this finished and see if anyone uses it and what the feedback is before we optimise?

Reply:
You are already saving this data in memory in class member `imageURL`.
So what you have to do is to avoid recomputing it as soon as `imageURL` is not null.
This could be done in initSession by calling getVehicle() only if car.getImageUrl() is null ?

Reply:
I am wrong because you are instantiating a new `MyRenaultHttpSession` at each refresh cycle and so with an empty `Car `structure.
So in this case, this is more the thing handler that should keep a trace (boolean) if you already got the imageUrl and if this is the case you should request `MyRenaultHttpSession` to not request `getVehicle`

Reply:
OK.  I will add an hasBeenInitialized flag or something to the handler and push a new version that only http gets what it need to.  Hope to get that done this week.  New features at this stage are maybe not great but it will improve the implementation to make it more efficient.